### PR TITLE
feat(token): create shared package to create and verify STS signed requests

### PR DESCRIPTION
**Pull request by dmlittle**
# What
- [x] Create `token` package used to generate and verify STS tokens

### CODE REVIEW

File: pkg/shared/token/token.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+	awsarn "github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/pkg/errors"
+)
+
+const (
+	pharosPrefix = "pharos-v1."
+	hostRegexp   = `^sts(\.[a-z1-9\-]+)?\.amazonaws\.com(\.cn)?$`
```
Comment (arkadiyt) [inline pkg/shared/token/token.go]: Can you kill the `.cn`?

File: pkg/shared/token/token.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+// GetSTSToken returns a token that contains a presigned AWS STS request.
+func (g generator) GetSTSToken() (string, error) {
+	request, _ := g.STSClient.GetCallerIdentityRequest(&sts.GetCallerIdentityInput{})
+
+	// Sign the request.  The expires parameter (sets the x-amz-expires header) is
+	// currently ignored by STS, and the token expires 15 minutes after the x-amz-date
+	// timestamp regardless.
+	// https://github.com/aws/aws-sdk-go/issues/2167
+	presignedURLString, err := request.Presign(60)
```
Comment (emilyzhang) [inline pkg/shared/token/token.go]: If the expire parameter is currently ignored, should we set it to 15 so if it is ever changed to not be ignored, our tokens will still expire in the same amount of time?

Comment (emilyzhang) [reply pkg/shared/token/token.go]: oh jk i read the definition for this functions, seems like it won't be changed for services that aren't S3, so disregard this

File: pkg/shared/token/token.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+			RequestID string `json:"RequestId"`
+		} `json:"ResponseMetadata"`
+	} `json:"GetCallerIdentityResponse"`
+}
+
+type tokenVerifier struct {
+	client *http.Client
+}
+
+// NewVerifier creates a Verifier that is able to verify the pharos tokens
```
Comment (emilyzhang) [inline pkg/shared/token/token.go]: super super super nit: add period to the end of this sentence

File: pkg/shared/token/token.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
+// Package token is heavily inspired in aws-iam-authenticator's token package.
```
Comment (emilyzhang) [inline pkg/shared/token/token.go]: super nit: "is heavily inspired by" and "pharos'" instead of "pharo's"

File: pkg/shared/token/token_test.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+	account = "123456789012"
+	userID  = "Alice"
+)
+
+type mockClient struct {
+	stsiface.STSAPI
+}
+
+func (m mockClient) GetCallerIdentityRequest(input *sts.GetCallerIdentityInput) (*request.Request, *sts.GetCallerIdentityOutput) {
+
```
Comment (emilyzhang) [inline pkg/shared/token/token_test.go]: super nit: remove line here

File: pkg/shared/token/token_test.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+func TestVerifyInvalidCanonicalARNError(t *testing.T) {
+	_, err := newVerifier(200, jsonResponse("arn", "1000", "userid"), nil).Verify(validToken)
+	errorContains(t, err, "arn 'arn' is invalid:")
+	assert.Error(t, err)
+}
+
+func TestVerifyInvalidUserIDError(t *testing.T) {
+	_, err := newVerifier(200, jsonResponse("arn:aws:iam::123456789012:user/Alice", "123456789012", "not:vailid:userid"), nil).Verify(validToken)
```
Comment (emilyzhang) [inline pkg/shared/token/token_test.go]: supersupernit: `not:valid:userid`

**Pull request merged**